### DISCUSSION TITLE
Read only mirror for LFS using only HTTP GET, akin to the dumb git HTTP protocol

### DISCUSSION THREAD

**Discussion by michalc** [git-lfs/git-lfs#5208] [Q&A] [answered]
I'm wanting to make a read only HTTP mirror of some git repositories, including files stored on LFS.

It looks fairly trivial to serve git repositories using a static server that supports only "static" GET HTTP requests - using the "dumb" protocol https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols#_dumb_http.

However, on the LFS side, so far it doesn't look like LFS has a similar "dumb" mode. From my initial investigations, it seems that at a minimum an endpoint that can handle POST batch requests is required.

Is there a way to have an LFS server where everything can be handled by static GET requests?

(Also aim here is to have a mirror powered only by S3, but that's an implementation detail I think)

**Comment by chrisd8088** [comment, marked as answer]
Yes, I believe that's correct.  As noted in [documentation](https://github.com/git-lfs/git-lfs/blob/13e6543151273578558449e3c60cde3fe40bc95b/docs/api/batch.md) of the Git LFS Batch API:

> All Batch API requests use the POST verb, and require the following HTTP headers. The request and response bodies are JSON.

This is because the Git LFS client has to send JSON to the server, and while it's perhaps not technically incorrect to do so with a `GET` request, it's definitely somewhat unconventional.

**Reply by michalc** [reply]
Ah thanks!

(Is it a shame in a way? The protocol _could_ be designed in such a way to support only static HTTP GET requests, much like the regular git dumb HTTP API?)

**Reply by chrisd8088** [reply]
I don't know that the question has ever come up previously, and the challenge, I think, is that assuming one is avoiding an API which depends on `GET` requests to have request bodies (i.e., JSON in this case), it would make the API more verbose.  Presumably fetching each object would require a `GET` to the API, to retrieve the actual URL for the object along with the necessary request headers, and then a request to fetch the object using that URL.  So instead of a single batch request to fetch multiple objects' URLs, you'd make a request to the API for each object to fetch its own URL.  Not terrible, of course, just more verbose.

**Reply by michalc** [reply]
Ah understood. I certainly agree it would be more verbose. Thanks!

(Also I guess for comment/posterity: this is my proof of concept server that sits in front of a dumb git http protocol server https://github.com/uktrade/git-lfs-http-mirror)

**Reply by bk2204** [reply]
There's a proposal for static data in #4193, but we haven't gotten around to implementing it yet, and it of course could change in the future.
